### PR TITLE
Prototype definition-based extraction

### PR DESCRIPTION
**Pull request by jingtang10**

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ * the same time relatively easy to implement.
+ *
+ * WARNING: This is not production-ready.
+ */
+internal object ResourceMapper {
+
+    /**
+     * Extract a FHIR resource from the `questionnaire` and `questionnaireResponse`.
+     *
+     * This method assumes there is only one FHIR resource to be extracted from the given
```
Comment (joiskash) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:51]: Just  putting this here so we can ask the Tuesday WG if this assumptions holds good most of the time.
> In some cases, the context of an element will switch. For example, an item might have a definition of "http://hl7.org/fhir/MedicationStatement#MedicationStatement.medicationReference" and a questionnaire-itemContext extension that shifts the context to "Medication". Subsequent children would then have definitions based on the "http://hl7.org/fhir/Medication" resource.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+     * `questionnaire` and `questionnaireResponse`.
+     */
+    fun extract(
+        questionnaire: Questionnaire,
+        questionnaireResponse: QuestionnaireResponse
+    ): Message {
+        val builder = questionnaire.itemContextNameToExpressionMap.values.first().let {
```
Comment (joiskash) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:58]: > Include the questionnaire-itemContext extension either on the Questionnaire root or on 'group' items within the Questionnaire to identify the resource that will serve as the context for any extraction.
Do we need a validator to make sure this is not violated

**Review by joiskash — CHANGES_REQUESTED** [review]
Great work! The code looks very neat! Just a few comments

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 1 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:51]: Created https://github.com/google/android-fhir/issues/240

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+/**
+ * Returns the field name for the [Questionnaire.Item]'s definition.
+ *
+ * For example, if the definition URI is
+ * `"http://hl7.org/fhir/StructureDefinition/Patient#Patient.birthDate"`, this function will return
+ * `"birthDate"`.
+ */
+private val Questionnaire.Item.definitionFieldName
+    get() = this.definition.value.substringAfterLast(".")
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:120]: Created https://github.com/google/android-fhir/issues/240

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+/**
+ * Extracts value for field in the builder from the corresponding question and answer in
+ * [questionnaireItem] and [questionnaireResponseItem].
+ */
+private fun Message.Builder.extractField(
+    questionnaireItem: Questionnaire.Item,
+    questionnaireResponseItem: QuestionnaireResponse.Item
+) {
+    val targetFieldName = questionnaireItem.definitionFieldName
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:99]: Created https://github.com/google/android-fhir/issues/240

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 1 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:51]: Cool. I think that's quite an advanced use case and happy to leave it to the future. But yes let's definitely discuss this on tuesday.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 2 above)
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:58]: I'm not sure we need a validator -- if the extension is not on the root or on a group we'll simply ignore it. But we do need to handle extensions on groups (which we don't). Again I feel this is quite advanced and I'm quite happy to leave it to later.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:101]: also added a note in the code to reference that issue.

**Comment by jingtang10** [comment]
Thanks @joiskash for the review. Please take another look.

**Review by joiskash — APPROVED** [review]
Ok great! The advance features can be added in the future!

**Pull request merged**